### PR TITLE
TalkNet preliminary part

### PR DESCRIPTION
**Pull request by stasbel**

### CODE REVIEW

File: nemo/backends/pytorch/nm.py [hunk 1]
```diff
@@ -106,7 +106,7 @@ def restore_from(self, path, local_rank=0):
             load_device = f"cuda:{local_rank}"
         else:
             load_device = self._device
-        self.load_state_dict(t.load(path, map_location=load_device))
+        self.load_state_dict(t.load(path, map_location=load_device), strict=False)
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/backends/pytorch/nm.py:109]: I was using this for unstrict loading for QN checkpoints (like, some layers fit and some don't). This has nothing to do with TN, so I revert this.

File: nemo/collections/asr/parts/features.py [hunk 2]
```diff
@@ -230,6 +230,8 @@ def forward(self, input_data):
                 )
         self.log_zero_guard_type = log_zero_guard_type
 
+        self.sample_rate, self.window_type, self.fmin, self.fmax = sample_rate, window, lowfreq, highfreq
+
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/asr/parts/features.py:234]: No.

File: nemo/backends/pytorch/actions.py [hunk 3]
```diff
@@ -526,6 +526,18 @@ def _eval(self, tensors_2_evaluate, callback, step, verbose=False):
                         tensors_list = []
                         # where we will all_gather tensor sizes
                         tensor_on_worker = registered_e_tensors[key]
+
+                        if not isinstance(tensor_on_worker, torch.Tensor):  # For string and other.
+                            if self.global_rank == 0:
+                                values_dict[key] = [tensor_on_worker] + ([None] * (world_size - 1))
+                            continue
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/backends/pytorch/actions.py:533]: Image I want to log something with raw python strings.
For example (my case): I use this for logging audio + raw text transcript (as transcription) for W&B.

File: nemo/utils/argparse.py [hunk 4]
```diff
@@ -53,13 +53,15 @@ def __init__(self, **kwargs):
         self.add_argument(
             "--train_dataset", type=str, default=None, help="training dataset path",
         )
+        self.add_argument('--eval_names', type=str, nargs="*", default=[], help="Eval datasets names.")
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/utils/argparse.py:56]: What would be instead?
Let me change it after NeMoArgParser removing. The is no point to do it now.

File: nemo/collections/tts/talknet_modules.py [hunk 5]
```diff
@@ -0,0 +1,916 @@
...
+        elif self._reduction == 'batch':
+            loss = loss.sum(-1) / mask.sum(-1)
+            loss = loss.mean()
+        else:
+            raise ValueError("Wrong reduction.")
+
+        return loss
+
+
+class WaveGlowInference:
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:876]: For complete TalkNet inferece notebook and logging to W&B with sampling.

[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:876]: Like, If I want to really hear how TalkNet sounds during training.

File: nemo/collections/tts/talknet_modules.py [hunk 6]
```diff
@@ -0,0 +1,916 @@
...
+            raise ValueError("Wrong Method")
+
+        dur_pred[dur_pred < 0.0] = 0.0
+        dur_pred = dur_pred.float().round().long()
+        tensors.dur_pred = dur_pred
+
+        return tensors
+
+
+class TalkNetMelsLoss(LossNM):
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:817]: `pad16` is true if we using padding to 16 trick for mel tensors.
`doubling` is trick of predicting 160 mels instead of 80 for 2x less time axis.

File: nemo/collections/tts/talknet_modules.py [hunk 7]
```diff
@@ -0,0 +1,916 @@
...
+
+        text_rep = torch.stack([b1[i : i + self._max_len] for i, b1 in zip(inds, text_rep)])
+        text_rep_mask = torch.stack([b1[i : i + self._max_len] for i, b1 in zip(inds, text_rep_mask)])
+        mel_true = torch.stack([b1[:, i : i + self._max_len] for i, b1 in zip(inds, mel_true)])
+        mel_len = torch.clamp(mel_len, max=self._max_len)
+
+        return text_rep, text_rep_mask, mel_true, mel_len
+
+
+class TalkNetDursLoss(LossNM):
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:657]: DONE

File: nemo/collections/tts/talknet_modules.py [hunk 8]
```diff
@@ -0,0 +1,916 @@
...
+        totals = self._generate_text_rep(dur, dur) + 1
+
+        left_c = (ones_cm - pos_cm) * mask
+        left_c = left_c.to(dtype=self._emb.weight.dtype)
+        left_c = 1.0 - left_c / totals  # noqa
+
+        return left_c
+
+
+class TalkNet(nemo_nm.TrainableNM):
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:506]: DONE

File: nemo/collections/tts/talknet_modules.py [hunk 9]
```diff
@@ -0,0 +1,916 @@
...
+        for b_i in b_indices:
+            yield from batches[b_i]
+
+
+class AllSampler(torch.utils.data.distributed.DistributedSampler):  # noqa
+    def __iter__(self):
+        return iter(list(range(len(self.dataset))))
+
+
+class BDAugs:
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:186]: OK

File: nemo/collections/tts/talknet_modules.py [hunk 10]
```diff
@@ -0,0 +1,916 @@
...
+    @property
+    def dataset(self) -> Optional[torch.utils.data.Dataset]:  # noqa
+        return None
+
+    @property
+    def data_iterator(self) -> Optional[torch.utils.data.DataLoader]:  # noqa
+        return self._dataloader
+
+
+class PolySpanEmb(nn.Module):
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:457]: DONE

File: nemo/collections/tts/talknet_modules.py [hunk 11]
```diff
@@ -0,0 +1,916 @@
...
+        if self._speaker_embs is not None:
+            example['speaker_emb'] = torch.tensor(self._speaker_embs[example['speaker']])
+
+        return example
+
+    def __len__(self):
+        return len(self._audio_dataset)
+
+
+class SuperSmartSampler(torch.utils.data.distributed.DistributedSampler):  # noqa
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:133]: DONE

File: nemo/collections/tts/talknet_modules.py [hunk 12]
```diff
@@ -0,0 +1,916 @@
...
+        return xy
+
+    @staticmethod
+    def pad(x, to=16, dim=1, value=0):
+        pads = [0, 0] * len(x.shape)
+        pads[-(2 * dim + 1)] = (to - (x.shape[dim] % to)) % to
+        return F.pad(x, pads, value=value)
+
+
+class TalkNetDataset:
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:85]: DONE

File: nemo/collections/tts/talknet_modules.py [hunk 13]
```diff
@@ -0,0 +1,916 @@
...
+    'TalkNetDataLayer',
+    'TalkNet',
+    'LenSampler',
+    'TalkNetDursLoss',
+    'TalkNetMelsLoss',
+    'WaveGlowInference',
+]
+
+
+class Ops:
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:54]: DONE

File: nemo/collections/tts/parts/talknet.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
+import numpy as np
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/parts/talknet.py:1]: DONE

File: nemo/collections/tts/parts/__init__.py [hunk 15]
```diff
@@ -11,6 +11,7 @@
 )
 from .layers import get_mask_from_lengths
 from .tacotron2 import Decoder, Encoder, Postnet
+from .talknet import dmld_loss, dmld_sample
 from .waveglow import WaveGlow
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/parts/__init__.py:15]: DONE

File: nemo/collections/tts/talknet_modules.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:817]: DONE

File: nemo/collections/asr/parts/dataset.py [hunk 16]
```diff
@@ -222,7 +228,18 @@ def __getitem__(self, index):
...
+        if self._add_text_raw:
+            output = (sample.text_raw,) + output
+
+        if self._add_id:
+            output = (sample.id,) + output
+
+        if self._add_speaker:
+            output = output + (sample.speaker,)
+
+        return output
```
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/asr/parts/dataset.py:242]: Don't you think they are common and could be reused somewhere?

**Comment by stasbel** [comment]
@blisc Rebased and problems fixed, your turn.

File: nemo/collections/asr/parts/dataset.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/asr/parts/dataset.py:242]: I'm sure they can be reused somewhere but if your code if going to be a list of if statements, I think it's worth thinking about why we should have this list of if statements. Is there a reason for it? Can you improve upon the design in some way that doesn't require a list of ifs?

File: nemo/collections/tts/parts/__init__.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/parts/__init__.py]: Thanks for changing the imports, can you also add the copyright header?

File: nemo/utils/argparse.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/utils/argparse.py:56]: I would revert all changes to this file.

File: nemo/backends/pytorch/actions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/backends/pytorch/actions.py:533]: How are you passing a raw python string inside of NeMo? Last I checked, NeMo does not support non-tensors

File: nemo/collections/tts/talknet_modules.py [hunk 17]
```diff
@@ -0,0 +1,916 @@
...
+            text_mask=NeuralType(('B', 'T'), MaskType()),
+            dur=NeuralType(('B', 'T'), LengthsType()),
+            text_rep=NeuralType(('B', 'T'), LengthsType()),
+            text_rep_mask=NeuralType(('B', 'T'), MaskType()),
+            text_raw=NeuralType(),
+            speaker=NeuralType(('B',), EmbeddedTextType(), optional=True),
+            speaker_emb=NeuralType(('B', 'T'), EncodedRepresentation(), optional=True),
+        )
+
+    def __init__(
```
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/talknet_modules.py:318]: This is for `TalkNetDataLayer`

File: nemo/collections/tts/talknet_modules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/talknet_modules.py:939]: Can you use the previous waveglow inference modules?

File: nemo/collections/tts/parts/__init__.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/parts/__init__.py]: There is a one-liner that I add. Replace with big one.

File: nemo/collections/tts/talknet_modules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:939]: You mean creating graphs and running neural_factory.infer and other stuff?
No, I don't think so. There is no point for this complication. I dont want WG graph to be part of TN computational graph.

File: nemo/backends/pytorch/actions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/backends/pytorch/actions.py:533]: It's actually does. The trick is `text_raw=NeuralType(),`.

File: nemo/utils/argparse.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/utils/argparse.py:56]: DONE
So, there we plan to have separate parser for every main file?

File: nemo/collections/tts/talknet_modules.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:318]: DONE

File: nemo/collections/asr/parts/dataset.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/asr/parts/dataset.py:242]: Modify a little.

**Comment by stasbel** [comment]
@blisc Your turn.

File: nemo/utils/argparse.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/utils/argparse.py]: At the current rate. None of nemo_nlp is using NeMoArgParser and no one is really consistent with it right now. So it seems to be a wasted effort.

File: nemo/backends/pytorch/actions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/backends/pytorch/actions.py:533]: @okuchaiev or @tkornuta-nvidia
Can you guys comment on this change? And make a decision on whether we want to make this change? I am conflicted about accepting this.

File: nemo/collections/tts/talknet_modules.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/talknet_modules.py:319]: Can you either add typehints in the code itself or in the __init__ docstring?
eg:
```
Args:
    data (str): Path to ...
```

File: nemo/collections/tts/talknet_modules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/talknet_modules.py:969]: Where do you use this class?

File: nemo/collections/asr/parts/dataset.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/asr/parts/dataset.py]: In my opinion, it seems much cleaner for you to create a new dataset that subclasses the original `AudioDataset` and returns what `AudioDataset` returns in addition to what you need. Something like
```
# Or some other meaningful name
class AudioTextDatasetWithSpeakerInformation(Dataset):
    def __init__(
        self,
        manifest_filepath,
        labels,
        featurizer,
        max_duration=None,
        min_duration=None,
        max_utts=0,
        blank_index=-1,
        unk_index=-1,
        normalize=True,
        trim=False,
        bos_id=None,
        eos_id=None,
        load_audio=True,
        parser='en',
    ):
        super().__init__(... All the params)

    def __getitem__(self, index):
        output = super().__getitem__(index)
        output.append(sample.id, sample.text_raw, sample.speaker)
        return output
```

But I will leave this decision to you. If you like your method better, I will accept it.

File: nemo/backends/pytorch/actions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/backends/pytorch/actions.py:533]: There is no other way if you want to log something with python objects (from my pov).

File: nemo/collections/tts/talknet_modules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:969]: In my dev notebooks and for my eval logger (which I deleted for this PR). This is still essential part for audio generating pipeline.

File: nemo/collections/asr/parts/dataset.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/asr/parts/dataset.py]: I am actually like idea of loading everything in single AuidoDataset and then use whatever you need in model's separate data loader a bit more. The only real speed-up is controlled by `load_audio` flag. I use `add_misc` just to the sake of compatibility with other pipelines.

File: nemo/collections/tts/talknet_modules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/talknet_modules.py:969]: Just remove this from this PR. In general, any class that has this comment `# One nasty little hack` will not be merged into master

File: nemo/backends/pytorch/actions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tkornuta-nvidia) [reply nemo/backends/pytorch/actions.py:533]: Hey, we have discussed that with @blisc and concluded that this will work. Actually I was discussing the use-cases with @ekmb  some time ago where she wanted to send strings (in SGD). As far as I remember she was using torch.distributed, it was raising an error, it seems to me that you just also solved her issue ;)

Thanks @stasbel 👍

[the remark this answers is not in the document]
Comment (tkornuta-nvidia) [reply nemo/backends/pytorch/actions.py:533]: Actually I think I got a use case that will not work here.

List of bounding boxes in ObjectDetection. I see that as list (batch) of list of 1D, 4-element tensors.

I think I got an idea (actually options) how to refine this, probably it should be discussed once I will be working on OD - which is phase III of CV collection development, and we have just passed stage I. So this is not going to happen soon.

That is why I still think it should be merged.

[the remark this answers is not in the document]
Comment (okuchaiev) [reply nemo/backends/pytorch/actions.py:533]: As far as NeuralTypes are concerned "string", "float", "float16", "int" are "storage" types which we don't enforce unless the user explicitly declares it in a way similar to how we declare sampling rate.
So, technically, you should be able to declare something like List(Tensor[string]). But this is only used to enforce validity of connections. We currently do not check if module's implementation actually obeys it's types.

Speaking of this particular change - it makes sense to me, so I'm OK for including this.

File: nemo/collections/tts/talknet_modules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:969]: DONE

File: nemo/collections/tts/talknet_modules.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (stasbel) [reply nemo/collections/tts/talknet_modules.py:319]: DONE

**Comment by stasbel** [comment]
@blisc DONE
